source/slang/slang-ir-inst-defs.h

```diff
@@ -507,14 +509,12 @@ INST(HighLevelDeclDecoration,               highLevelDecl,          1, 0)
 
     INST(BindExistentialSlotsDecoration, bindExistentialSlots, 0, 0)
 
-        /// A `[polymorphic]` decoration marks a function parameter that should translate to an abstract type
-        /// e.g. (void*) that are casted to actual type before use. For example, a parameter of generic type
-        /// is marked `[polymorphic]`, so that the code gen logic can emit it as a `void*` parameter,
-        /// allowing the function to be used at sites that are agnostic of the actual object type.
-    INST(PolymorphicDecoration, polymorphic, 0, 0)
 
         /// A `[this_ptr]` decoration marks a function parameter that serves as `this` pointer.
-    INST(ThisPointerDecoration, this_ptr, 0, 0)
+        /// `[this_ptr]` decoration is also used to mark an `IRFunc` as a non-static function.
+        /// The argument is an integer value that represents the index of the `this` parameter,
+        /// which is always 0.
+    INST(ThisPointerDecoration, this_ptr, 1, 0)
```

Comment:
We could have all `this` typed parameters to have `IRThisType` instead of using this decoration. But I would like to do that in another PR instead.

Reply:
There is a design question here of how `IRThisType` would be used in the encoding. I can see two big options:

1. There is a single global `IRThisType` instance, and all interfaces just use it as a placeholder in their requirement type signatures for wherever the `this` paramter, or another parameter of type `This` goes.

2. Each `IRInterfaceType` `I` has an associated `ThisType<I>` that is specific to it. This allows us to know the interface that each `This`-typed parameter was associated with, in case that matters.

Option (2) potentially creates a circular-reference situation (the interface includes operands that reference the `ThisType` while the type references the interface). That circularity would be easier to navigate in the case where we change the encoding of interfaces so that they are a parent instruction and the entries are child instructions (similar to how witness tables are encoded at the moment).

Reply:
Done. Option 1 is implemented.